fix performance regressions introduced in #417

o- opened the pull request:
some part of the scope analysis state can be queried at any position and only the actual bindings are flow-sensitive.

o- committed:
fix performance regressions introduced in #417

charig commented on rir/src/compiler/analysis/scope.h:
```diff
@@ -130,26 +130,49 @@ class ScopeAnalysis : public StaticAnalysis<
-        lookup(envs, i, [&](AbstractLoad load) { action(load.result); },
-               notFound);
+    //
+    // global version, can be called if the state does not correspond with i
```
value?

charig commented on rir/src/compiler/analysis/scope.cpp:
```diff
@@ -12,48 +12,72 @@ ScopeAnalysis::ScopeAnalysis(ClosureVersion* cls, Promise* prom, Value* promEnv,
-        action(superLoad(state, sld->varName, sld->env()));
-        return;
-    }
+    lookupGlobal(state, instr, action, [&]() {
```
Lookup calls lookupGlobal. So...I can't find the semantics difference between calling `lookupAt` and `lookupGlobal` in terms of the correspondence between the state and the current instruction

o- replied:
`lookupGlobal <= lookupAt`

something that is valiad at a particular position, is also valid globally, lookupAt also does lookupGlobal

charig replied:
Ok, I got it now, the issue is that you are defining the local lookup in the `not found lambda` of the `lookupGlobal` call. I still do not get use to this code style. Just random comment addition: I know we discussed this since its introduction and I can not come to a better alternative but lookup is pretty confusing in this context.

o- committed:
extend static analysis to have flow-insensitive global state and use it for scope analysis

In the scope analysis we need to progressively build up the analysis
results, to be able to resolve recursive lookups. This commit makes this
explicit, in keeping a global store for results than can be queried
anytime.

o- commented:
@charig can you take another look at my newest commit? I found another issue that was not fixed in #417, which caused more problems. So I had to slightly revise the strategy.

o- committed:
avoid double counting introduced in #437

fix for the fix...
